libimage/runtime.go

```diff
@@ -565,7 +573,42 @@ func (r *Runtime) ListImages(ctx context.Context, names []string, options *ListI
 		}
 	}
 
-	return r.filterImages(ctx, images, options)
+	filtered, err := r.filterImages(ctx, images, options)
+	if err != nil {
+		return nil, err
+	}
+
+	if !(options.Dangling || options.Parent) {
+		return filtered, nil
+	}
+
+	// If explicitly requested by the user, pre-compute and cache the
+	// dangling and parent information of all filtered images.  That will
+	// considerably speed things up for callers who need this information
+	// as the layer tree will computed once for all instead of once for
+	// each individual image (see containers/podman/issues/17828).
+
+	tree, err := r.layerTree()
```

Comment:
Note that `filterImages` can also trigger a `layerTree` computation.

I suppose a clean way to deal with that could be to have a separate `lazyLayerTree` holder, passed to the various callees of of `ListImages` (and to things like `getChildren`?). That could also avoid the need for the `if !(options.Dangling || options.Parent)` above.

Asymptotically, calling `layerTree` once vs. twice makes no difference, and it’s a much smaller performance win than the primary purpose of this PR, so I’m absolutely fine with not dealing with the two calls in _this_ PR. Maybe just a comment to note this duplication?